api/s2n.h

```diff
@@ -687,6 +687,269 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ */
+S2N_API 
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+
+/**
+ * Appends a PSK object to the list of PSKs supported by the s2n connection. 
+ * If a PSK with a duplicate identity is found, an error is returned and the PSK is not added to the list.
+ * Note that the PSK object is copied to the connection and the user is responsbile for freeing the memory associated with the s2n_psk object. 
+ *
+ * @param conn A pointer to the s2n_connection object that contains the list of PSKs supported.
+ * @param psk A pointer to the `s2n_psk` object to be appended to the list of PSKs on the s2n connection.
+ */
+S2N_API 
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+
+/**
+ * The list of PSK modes supported by s2n-tls for TLS versions >= TLS1.3.
+ * Currently s2n-tls supports two modes - `S2N_PSK_MODE_RESUMPTION`, which represents the PSKs established 
+ * using the previous connection via session resumption, and `S2N_PSK_MODE_EXTERNAL`, which represents PSKs 
+ * established out-of-band/externally using a secure mutually agreed upon mechanism.
+ * Note that currently s2n-tls supports only PSK with ECDHE key exchange, which provides forward secrecy.
+ */ 
+typedef enum {
+    S2N_PSK_MODE_RESUMPTION,
+    S2N_PSK_MODE_EXTERNAL 
+} s2n_psk_mode;
+
+/**
+ * Sets the PSK mode on the s2n config object. 
+ * The supported PSK modes are listed in the enum `s2n_psk_mode` above. 
+ * 
+ * @param config A pointer to the s2n_config object being updated.
+ * @param mode The PSK mode to be set. The supported psk modes are listed in the enum `s2n_psk_mode` above. 
+ */
+S2N_API 
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+
+/**
+ * Sets the PSK mode on the s2n connection object.
+ * The supported PSK modes are listed in the enum `s2n_psk_mode` above. 
+ * This API overrides the PSK mode for the connection and will continue to use the existing 
+ * PSK mode even when a new config is set for the connection.
+ * 
+ * @param conn A pointer to the s2n_connection object being updated.
+ * @param mode The PSK mode to be set.
+ */
+S2N_API 
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+
+/**
+ * Gets the negotiated PSK identity length from the s2n connection object. The negotiated PSK 
+ * refers to the chosen PSK by the server to be used for the PSK connection. 
+ * If the negotiated PSK does not exist, the value `0` is returned.  
+ * 
+ * @param conn A pointer to the s2n_connection object that successfully negotiated a PSK connection.
+ * @param identity_length The length of the negotiated PSK identity. 
+ */
+S2N_API 
+int s2n_connection_get_negotiated_psk_identity_length(struct s2n_connection *conn, uint16_t *identity_length);
+
+/**
+ * Gets the negotiated PSK identity from the s2n connection object. 
+ * If the negotiated PSK does not exist, the PSK identity will not be obtained and no error will be returned. 
```

Comment:
how do we indicate an error/PSK does not exist? is that via the previous call to get length?

Reply:
Yes, If negotiated PSK does not exist, identity will be not be obtained and should be `NULL`.  When negotiated PSK does not exist a call to `s2n_connection_get_negotiated_psk_identity_length` will return `0` as described in the previous call.